speculos/client.py

```diff
@@ -0,0 +1,181 @@
+    return data[:-2], status
+
+
+def screenshot_equal(path1: str, path2: str) -> bool:
+    """Compare two images and return True if they are equal."""
+
+    with Image.open(path1) as img1:
+        with Image.open(path2) as img2:
+            diff_img = ImageChops.difference(img1, img2)
+    return diff_img.getbbox() is None
+
+
+class ApduException(Exception):
+    def __init__(self, sw=0x6F00) -> None:
+        self.sw = sw
+
+    def __str__(self) -> str:
+        return f"Exception: invalid status 0x{self.sw:x}"
+
+
+class ApduResponse:
+    def __init__(self, response: requests.Response) -> None:
+        self.response = response
+
+    def receive(self) -> bytes:
+        assert self.response.status_code == 200
+        data, status = split_apdu(bytes.fromhex(self.response.json()["data"]))
+        if status != 0x9000:
+            raise ApduException(status)
+        return data
+
+
+class Api:
+    def __init__(self, api_url: str) -> None:
+        self.api_url = api_url
+        self.timeout = 2000
+        self.session = requests.session()
+        self.stream = self._open_stream()
+
+    def _open_stream(self) -> requests.Response:
+        stream = self.session.get(f"{self.api_url}/events?stream=true", stream=True)
+        assert stream.status_code == 200
+        return stream
+
+    def get_next_event(self):
+        line = self.stream.raw.readline()
+        event = json.loads(line)
+        assert event
+        return event
+
+    def wait_for_text_event(self, text: str) -> dict:
+        """Wait until an event containing the specified text is received."""
+
+        while True:
+            event = self.get_next_event()
+            if text in event["text"]:
+                break
+        return event
+
+    def press_and_release(self, button: str):
+        assert button in ["left", "right", "both"]
+        data = {"action": "press-and-release"}
+        with self.session.post(f"{self.api_url}/button/{button}", json=data) as response:
+            assert response.status_code == 200
+
+    def finger_touch(self, x: int, y: int):
+        data = {"action": "press-and-release", "x": x, "y": y}
+        with self.session.post(f"{self.api_url}/finger", json=data) as response:
+            assert response.status_code == 200
+
+    def get_screenshot(self) -> bytes:
+        with self.session.get(f"{self.api_url}/screenshot") as response:
+            assert response.status_code == 200
+            assert response.headers["Content-Type"] == "image/png"
+            return response.content
+
+    def _apdu_exchange(self, data: bytes) -> bytes:
+        with self.session.post(f"{self.api_url}/apdu", json={"data": data.hex()}) as response:
+            apdu_response = ApduResponse(response)
+            return apdu_response.receive()
+
+    def _apdu_exchange_async(self, data: bytes) -> requests.Response:
+        return self.session.post(f"{self.api_url}/apdu", json={"data": data.hex()}, stream=True)
```

Comment:
"async" is misleading, in the context of Python: `async` has a specific meaning, related to AsyncIO (https://docs.python.org/3/library/asyncio.html).

I suggests other names: `apdu_exchange_nopoll`, `apdu_exchange_nowait`, `apdu_exchange_request`